zebra-chain/src/sprout/tests/tree.rs

```diff
@@ -0,0 +1,45 @@
+//! Tests for Sprout Note Commitment Trees.
+
+use hex::FromHex;
+
+use crate::sprout::commitment::NoteCommitment;
+use crate::sprout::tests::test_vectors;
+use crate::sprout::tree;
+
+#[test]
+fn empty_roots() {
+    zebra_test::init();
+
+    for i in 0..tree::EMPTY_ROOTS.len() {
+        assert_eq!(
+            hex::encode(tree::EMPTY_ROOTS[i]),
+            test_vectors::HEX_EMPTY_ROOTS[i]
+        );
+    }
+}
+
+#[test]
+fn incremental_roots() {
+    zebra_test::init();
+
+    let mut leaves = vec![];
+
+    let mut incremental_tree = tree::NoteCommitmentTree::default();
+
+    for (i, cm) in test_vectors::COMMITMENTS.iter().enumerate() {
+        let bytes = <[u8; 32]>::from_hex(cm).unwrap();
+
+        let cm = NoteCommitment::from(bytes);
+
+        leaves.push(cm);
+
+        let _ = incremental_tree.append(cm);
+
+        assert_eq!(hex::encode(incremental_tree.hash()), test_vectors::ROOTS[i]);
```

Comment:
I can't figure out why I can't get the hashes to match here.

Reply:
Is it byte-reversed?

(Sometimes Zcash prints things in little-endian order, and sometimes in big-endian order.)

Reply:
It is. I've also tried non-byte-reversed vectors.

Reply:
I mean, the test vectors are byte-reversed.

Reply:
I think it's the level vs layer issue, which got me pretty confused as well. If you look at the Sapling code, in `EMPTY_ROOTS` the `[0]` element is the root and the last is the "leaf". In the Sprout code is the other way around. This means that the Sprout `Hashable::empty_root()` method  is returning the incorrect value (since it converts from `level`  to `layer` before accessing `EMPTY_ROOTS`). You can either invert `EMPTY_ROOTS` (building it in the reverse order, like the Sapling code), or change `empty_root()`  to not convert `level`  to `layer`.

(I haven't confirmed this, but it seems likely to be the issue)

Reply:
Thank you for the suggestion @conradoplg. That was exactly the issue. I fixed it by reversing the vector of empty roots. This solution is compatible with the Sapling implementation.